pkg/ottl/ottlfuncs/func_len.go

```diff
@@ -0,0 +1,39 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type LenArguments[K any] struct {
+	Target ottl.StringGetter[K] `ottlarg:"0"`
```

Comment:
My gut says that this function should be able to return the length of a slice and a pcommon.Slice as well. I think we'll should use a regular Getter  and do the type inference ourselves and return an error if the input is not a string, []any, or pcommon.Slice.

Reply:
We should consider adding Golang maps and pcommon.Map to that as well. Not sure what the immediate use cases would be, but it seems like something I'd expect if we have the other two.

Reply:
Ok I added support for this. My go kung fu is weak and I wasn't sure there was a better way than `reflect.ValueOf()` to determine if an `interface{}` is a generic slice, so please let me know if there's a better way.

Reply:
The standard way to do this is to have a type switch statement. Not sure how the speed differs between that approach and using reflection. You can see an example of how we're doing it for all supported OTTL data types here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/internal/value.go#L16.

Reply:
Also, in terms of supported types, I think we should mirror the [supported types](https://pkg.go.dev/builtin#len) for the len function with both Golang native types and the pdata analogs for those types.

Reply:
I believe the exact type list comes out to:

- string
- map[string]any
- []any
- pcommon.Map
- pcommon.Slice
- pcommon.Value of types String, Map, and Slice

Reply:
@evan-bradley I don't think type switch statement can be used to match a generic slice. See this [stackoverflow post](https://stackoverflow.com/questions/38748098/golang-type-switch-how-to-match-a-generic-slice-array-map-chan) which reflects my experience.

So I think I have to use reflection for `string`, `map[string]any`, and `[]any`, but a type switch for `pcommon.Map`, `pcommon.Slice`, and `pcommon.Value` since those aren't in available to match in the `reflect` package.

Reply:
Do we need to support generic slices? So far in OTTL we've been able to take advantage of the fact that we can enumerate all the types we support because pdata only supports a limited set of types. See [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/internal/value.go#L16) for an example. Unfortunately it's a little verbose, but my suspicion is that it's faster than reflection. If it's not faster than reflection then I would be fine sticking with reflection. I think it should be possible to switch on pdata types with reflection using something similar to what we do [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/functions.go#L178).

Reply:
Its definitely possible to do a type switch on pcommon types if we're don't need to support generic map / slice.

I'll defer to the go experts on perf of reflection vs. type switch, and on whether we need to support generic map / slice vs. only pcommon versions.

Reply:
@evan-bradley we need to be able to handle generic slices because the language allows stuff like: `set(attributes["len"], Len(["a", "slice"]))`.  It is silly, but doable.

Reply:
@jack-berg I didn't test this but I think its the idea we want:

```go
switch v := val.(type) {
  case pcommon.Value:
    switch v.Type() {
      case pcommon.ValueTypeMap:
        v.Map().Len()
      case pcommon.ValueTypeSlice:
        v.Slice().Len()
      case pcommon.ValueTypeStr:
        len(v.Str())
      default:
        err
    }
  case pcommon.Slice:
    v.Len()
  case pcommon.Map:
    v.Len()
  default:
	  x := reflect.ValueOf(val)
	  switch x.Kind() {
	    case reflect.String, reflect.Slice, reflect.Map:
              x.Len()
            default:
	      err
	  }
  }
```

Reply:
> ["a", "slice"] will be a []Getter

good point, I forgot thats how we dealt with that.

Reply:
I did some very quick benchmark testing:

```
❯ go test -bench=. -run=notests --tags="" -benchmem github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs
BenchmarkReflectionMap-10       265206580                4.362 ns/op           0 B/op          0 allocs/op
BenchmarkInferenceMap-10        591525532                2.054 ns/op           0 B/op          0 allocs/op
BenchmarkReflectionSlice-10     549845289                2.182 ns/op           0 B/op          0 allocs/op
BenchmarkInferenceSlice-10      588511636                2.035 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs    6.180s

```

```go

func BenchmarkReflectionMap(b *testing.B) {
	var input any
	input = map[string]any{}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = testReflection(input)
	}
}

func BenchmarkInferenceMap(b *testing.B) {
	var input any
	input = map[string]any{}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = testInference(input)
	}
}

func BenchmarkReflectionSlice(b *testing.B) {
	var input any
	input = []int64{}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = testReflection(input)
	}
}

func BenchmarkInferenceSlice(b *testing.B) {
	var input any
	input = []int64{}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = testInference(input)
	}
}

func testInference(input any) (int, error) {
	switch v := input.(type) {
	case map[string]any:
		return len(v), nil
	case []int64:
		return len(v), nil
	default:
		return 0, fmt.Errorf("invalid")
	}
}

func testReflection(input any) (int, error) {
	x := reflect.ValueOf(input)
	switch x.Kind() {
	case reflect.Slice, reflect.Map:
		return x.Len(), nil
	default:
		return 0, fmt.Errorf("invalid")
	}
}
```

@evan-bradley I don't think we can enough improvement to justify the sprawl.

Reply:
Thank you for testing that. Honestly it's a relief since it's less to maintain, let's stick with reflection.

Reply:
Cool. Swamped tonight but will push up a commit in the morning with the (hopefully) last round of updates. Feel free to add to this PR if its needed sooner for some reason.

Reply:
No problem.  Thanks for taking this on, its super important!